proposals/052-cryptography-leg-1.md

```diff
@@ -0,0 +1,329 @@
+
+- [Haskell Cryptography Group ](https://github.com/haskell-cryptography)
+
+As the intended owner of the project, this project will be new maintenance burden on the Haskell Cryptography Group.
+
+- Users of existing cryptography libraries such as `crypton`, `cryptonite`, `libsodium / saltine`.
+
+As the intended target audience of this project, this proposal aims to directly benefit developers by providing high-quality cryptography libraries at several appropriate levels of abstraction, but will also place the burden of migration to these new libraries on developers.
+
+- Downstream users of Haskell software, developers with 
+
+As the eventual user of the developed software, their safety and security will rest on the proper execution of this proposal. 
+
+## Risks
+
+There are a number of risks.
+
+- Cryptography is hard and each algorithm may have individual nuances or uncommon requirements. 
+- Botan FFI is incompletely documented and has gaps (stream ciphers, X509 stores, a few broken functions) which need to be fixed with C++ shims
+- Strictness, memory safety & scrubbing, C++ shims will take an unknown amount of time
+- A thorough / safe implementation will require consulting the C++ source of the FFI
+- Polynomial combinations of algorithms makes exhaustive testing complicated.
+
+These risks are amortized in that while they are tedious and potentially time-consuming, they may be considered extended goals of high value, and there is still substantial value that can be delivered without them, and the completion of risk items may continue in a future proposal leg.
+
+- Single point of failure
+
+This is a proposal for one full-time engineer; this constitutes a single point of failure. To combat this, the official github repo will be owned by the Haskell Cryptography Group, and another member of the Haskell Cryptography Group will be selected to hold backup keys and permissions to the website, server, and official repositories as needed. In the case of an emergency for which I am unreachable, this will allow for others to take over as necessary.
+
+## Success
+
+This leg of the project will be considered a success if it results in the submission of a candidate or accepted package to hackage, with production-ready bindings, documentation, tutorials, and unit tests as appropriate for the following botan-related libraries:
+
+- botan-bindings
+- botan-low
+- botan
+
+The following Botan library features are considered optional to the success of this leg, due to requirement of C++ / FFI shims or additional research which will take an unknown amount of work, but effort will be made to include them if possible within the timeframe. 
+
+- Extended X509 support
+- Stream ciphers
+- Test vectors for algorithms, especially [CAVP / FIPS / NIST-approved](https://csrc.nist.gov/projects/cryptographic-algorithm-validation-program)
```

Comment:
This seems like something that could be extremely valuable in its own right. What if you create a general lib for testing crypto, and then use it in your test suite and also make it available for other tests?

Reply:
Yes - the value of test suites has been discussed and recognized elsewhere in this PR, and will be emphasized.